File: app/src/main/java/com/bugsnag/android/SessionStore.java
package com.bugsnag.android;

import com.bugsnag.android.internal.ImmutableConfig;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.Calendar;
import java.util.Comparator;
import java.util.Date;
import java.util.UUID;

/**
 * Store and flush Sessions which couldn't be sent immediately due to
 * lack of network connectivity.
 */
class SessionStore extends FileStore {

    static final Comparator<File> SESSION_COMPARATOR = new Comparator<File>() {
        @Override
        public int compare(File lhs, File rhs) {
            if (lhs == null && rhs == null) {
                return 0;
            }
            if (lhs == null) {
                return 1;
            }
            if (rhs == null) {
                return -1;
            }
            String lhsName = lhs.getName();
            String rhsName = rhs.getName();
            return lhsName.compareTo(rhsName);
        }
    };

    SessionStore(@NonNull ImmutableConfig config,
                 @NonNull Logger logger,
                 @Nullable Delegate delegate) {
        super(new File(config.getPersistenceDirectory().getValue(), "bugsnag-sessions"),
                config.getMaxPersistedSessions(),
                SESSION_COMPARATOR,
                logger,
                delegate);
    }

    @NonNull
    @Override
    String getFilename(Object object) {
        return SessionFilenameInfo.defaultFilename();
    }

    public boolean isTooOld(File file) {
        Calendar cal = Calendar.getInstance();
        cal.add(Calendar.DATE, -60);
        return SessionFilenameInfo.findTimestampInFilename(file) < cal.getTimeInMillis();
    }

    public Date getCreationDate(File file) {
        return new Date(SessionFilenameInfo.findTimestampInFilename(file));
    }
}

File: app/src/main/java/com/bugsnag/android/SessionStore.kt
package com.bugsnag.android

import com.bugsnag.android.SessionFilenameInfo.Companion.defaultFilename
import com.bugsnag.android.SessionFilenameInfo.Companion.findTimestampInFilename
import java.io.File
import java.util.Calendar
import java.util.Comparator
import java.util.Date

/**
 * Store and flush Sessions which couldn't be sent immediately due to
 * lack of network connectivity.
 */
internal class SessionStore(
    bugsnagDir: File,
    maxPersistedSessions: Int,
    private val apiKey: String,
    logger: Logger,
    delegate: Delegate?
) : FileStore(
    File(bugsnagDir, "sessions"),
    maxPersistedSessions,
    logger,
    delegate
) {
    fun isTooOld(file: File?): Boolean {
        val cal = Calendar.getInstance()
        cal.add(Calendar.DATE, -60)
        return findTimestampInFilename(file!!) < cal.timeInMillis
    }

    fun getCreationDate(file: File?): Date {
        return Date(findTimestampInFilename(file!!))
    }

    companion object {
        val SESSION_COMPARATOR: Comparator<in File?> = Comparator { lhs, rhs ->
            if (lhs == null && rhs == null) {
                return@Comparator 0
            }
            if (lhs == null) {
                return@Comparator 1
            }
            if (rhs == null) {
                return@Comparator -1
            }
            val lhsName = lhs.name
            val rhsName = rhs.name
            lhsName.compareTo(rhsName)
        }
    }

    override fun getFilename(obj: Any?): String {
        val sessionInfo = defaultFilename(obj, apiKey)
        return sessionInfo.encode()
    }
}
